src/cluster_legacy.c

```diff
@@ -2146,12 +2146,21 @@ void clearNodeFailureIfNeeded(clusterNode *node) {
         clusterDoBeforeSleep(CLUSTER_TODO_UPDATE_STATE | CLUSTER_TODO_SAVE_CONFIG);
     }
 
+    /* If any of the replica of a given primary can't failover, then immediately mark it as alive. */
+    int cant_failover = 1;
+    for (int j = 0; j < node->num_replicas; j++) {
+        if (!clusterNodeIsNoFailover(node->replicas[j])) {
+            cant_failover = 0;
+            break;
+        }
+    }
+
     /* If it is a primary and...
      * 1) The FAIL state is old enough.
      * 2) It is yet serving slots from our point of view (not failed over).
      * Apparently no one is going to fix these slots, clear the FAIL flag. */
     if (clusterNodeIsVotingPrimary(node) &&
-        (now - node->fail_time) > (server.cluster_node_timeout * CLUSTER_FAIL_UNDO_TIME_MULT)) {
+        ((now - node->fail_time) > (server.cluster_node_timeout * CLUSTER_FAIL_UNDO_TIME_MULT) || cant_failover)) {
```

Comment:
`cant_failover` is not a good name. maybe `dont_wait`? or another option would be setting `node->fail_time` to 0 when we know no known replicas would like to step up.

PS, it would be more readable I think if this function returns after handling the non primary case. There is no overlapping between the handling of non primary and primary cases.

Reply:
You read my mind :). I was unhappy with the naming as well. `dont_wait` sounds better.